server/src/main/java/org/elasticsearch/search/aggregations/timeseries/TimeSeriesIndexSearcher.java

```diff
@@ -106,16 +116,27 @@ private boolean populateQueue(List<LeafWalker> leafWalkers, PriorityQueue<LeafWa
                 queue.add(leafWalker);
             }
         }
+        assert queueAllHaveTsid(queue, currentTsid);
+        // If all walkers are exhausted then nothing will have been added to the queue
+        // and we're done
         return queue.size() > 0;
     }
 
+    private boolean queueAllHaveTsid(PriorityQueue<LeafWalker> queue, BytesRef tsid) throws IOException {
+        for (LeafWalker leafWalker : queue) {
+            BytesRef walkerId = leafWalker.tsids.lookupOrd(leafWalker.tsids.ordValue());
+            assert walkerId.equals(tsid) : tsid.utf8ToString() + " != " + walkerId.utf8ToString();
+        }
+        return true;
+    }
+
     private static class LeafWalker {
         private final LeafCollector collector;
         private final Bits liveDocs;
         private final DocIdSetIterator iterator;
         private final SortedDocValues tsids;
-        private final SortedNumericDocValues timestamps;
-        int docId;
+        private final SortedNumericDocValues timestamps;    // TODO can we have this just a NumericDocValues?
```

Comment:
`@timestamp` is just a regular date at this point so it supports multiple values. I think it might make sense to build a special mapper for it, yeah. It'd be useful to:
1. force it to be single valued
2. test that it is within the timestamp range - right now we do this with the data stream mapper. that feels *weird* but it made sense at the time
3. store the timestamp in the parse context so we can grab it more easily. Right now we dig it out of the fields which are not stored in a kind way.
4. Index it with a `NumbericDocValuesField`

Reply:
This is not tracked in https://github.com/elastic/elasticsearch/issues/83604 and I think I would prefer to address this as a follow up.